### DISCUSSION TITLE
Configure fail2ban completly without firewall. Are there examples available?

### DISCUSSION THREAD

**Discussion by ftasnetamot** [fail2ban/fail2ban#3796] [Q&A]
Hello,

My plan is, setting up fail2ban without any firewall interaction.  Just make fail2ban update its database with informations pulled from logs, keep the ip addresses in the databases and delete them, when the bantime is over.  Just like it does that, with the usual iptables/nft interaction.
In addition, I like to configure my applications, giving informations, why and how long you are blocked.
That is relative easy: In exim you can do this like for dnsblock-lists, instead query a nameserver, do a sqlite query.
Similar to apache, with mod_rewrite RewriteMap will do the job
similar things can be done in dovecot and other programs.

This idea started, as I got problems with several setups, where blocking email via firewall gives a big legal risk because of  suppressing information. 

This should not start a flamewar, telling me, that only firewall blocks will keep the evil out.  I think, that a good setting may use both methods, depending on the threat level.

I have done some search, if anybody has already done something like this, and if example-configurations to start with, are somewhere available.

Thanks
f.

**Comment by sebres** [comment]
Basically, it is enough to overwrite `action` (or `banaction`) with some other non-firewall action of your choice (or with empty value, like `action =` if you don't want any action at all for some reasons). The actions may be shell actions or pythonic actions, so you can write them by yourself using some action from [action.d](https://github.com/fail2ban/fail2ban/tree/master/config/action.d) folder as an example.

So for instance in this jail nothing will be banned at all at the net-filter level:
```ini
[some-jail]
action =
enabled = true
```

Because you didn't describe your approach more concrete (to be quite honest, I catched only the part where you don't want to use banaction triggering something in firewall or net-filter, but not what you really want), maybe you'd find some clues in [this FAQ in our wiki](https://github.com/fail2ban/fail2ban/wiki/How-to-ban-something-other-as-host-(IP-address),-like-user-or-mail,-etc.).

However, since you mentioned the apache, as an example of nginx-based action you can take a look into [action.d/nginx-block-map.conf](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/nginx-block-map.conf), which can be used to ban directly in nginx by something like user, session or other captured data (even some combinations of them).
For the example how to use or configure it see in first large block of the action:
https://github.com/fail2ban/fail2ban/blob/766d2b8d742cf207f3321b27bf74fce2760a8eb7/config/action.d/nginx-block-map.conf#L5-L75

For the list of all tags, available in action, see [man jail.conf.5](https://manpages.debian.org/stable/fail2ban/jail.conf.5.en.html#Action_Tags) or
https://github.com/fail2ban/fail2ban/blob/766d2b8d742cf207f3321b27bf74fce2760a8eb7/fail2ban/server/actions.py#L374-L401

**Comment by ftasnetamot** [comment]
Ok, thanks for this answer.
How can I suppress the warning, that the jail has no actions?  

Ok, now I am a big step further with my exim solution.
A jail with a very cpu friendly filter<br>`failregex = ^%(pid)s .*?[^\(]\[<ADDR>\][^\)]`<br>will grep all lines from exim _rejectlog_ and not as most exim examples the _mainlog_.  In the _rejectlog_ I have everything exims has rejected.  This are in most cases those guys, I wish to block for long, but there may also -depending on the system- true users, having auth-problems while setting up new devices.  So the ban-conditions should set in a way, that those users get a chance to fix their problems.

The ACL in exim looks like this:
```
drop
    message     = You are blocked from this MX, because to many failed connection in the last 1h. Come back after fixing your problems and a grace period of 70 minutes
    log_message = fail2ban caused lock
    hosts       = ${lookup sqlite{/var/lib/fail2ban/fail2ban.sqlite3 select ip from bans where ip="$sender_host_address" AND jail="exim-rejectlog" ;}}
    delay       = 3s
```
Give the exim user (in Debian derived systems Debian-exim) additional read access to the sqlite Database file.
```
setfacl -Rdm u:Debian-exim:rX /var/lib/fail2ban
setfacl -Rm u:Debian-exim:rX /var/lib/fail2ban
setfacl -m u:Debian-exim:r /var/lib/fail2ban/fail2ban.sqlite3
```
The first option makes sure, that all newly created files under this directoy, will inherit those ACLs. The uppercase X gives directories an X-ACL, and files only, if there is an x on others.

Now I am going to some other applications.

**Reply by sebres** [reply]
> How can I suppress the warning, that the jail has no actions?

At the moment no way to do that excepting to increase log-level (but then you'd not see NOTICE level too, so no Ban/Unban messages etc).

You could simply write an action filling some lookup database for the exim (add something by ban and remove it by unban or some timeout), in this case it'd be a bit more consistent, I guess.

**Comment by ftasnetamot** [comment]
**Next question:**  I try to understand your nginx example as nginx offers dnsblocklist possibilities only in the paid + version :-(
Is the mapfile recreated, when single ips get unbanned? What happens, when the file gets recreated, and nginx will access it?
And I am still searching for a method, to suppress the warning, that a jails has no banaction.

I thought initially about having fail2ban feeding an local dynamic dns-zone with the banned ips, having the ban-action setting the RR and unban deleting the RR. But nginx offers dnsbl/rbl only on the paid + version :-(

However, I will give my private dns-blocklist a try, as this would make the ban-information available to a few more applications.
Its also thinkable, that with this way, multiple applications can be reached, as rbl/dnsbl is widespread. It is also thinkable, that with this method, applications can be protected from firewalls/proxies in front of the host, as long, as the private dns server is reachable from there.

While writing this, I start thinking. As this is "discussions", I start with a first idea:

Give the fact, IP 1.2.3.4 should be banned:
- banaction will set an DNS IP RR  4.3.2.1.blacklist.private.local pointing to: 
    - 127.0.0.1 if jail is exim
    - 127.0.0.2 if jail is apache
    - 127.0.0.3 if jail is ...
just like all lists work, definitions, which ip is which jail is totally free
- it will also set an DNS TXT RR 4.3.2.1.jailname.blacklist.private.local
  containing the reason, why it is blocked.

the additional subdomain is needed, because we need a reason for every jail.

Doing so, the unban-action can delete both RRs, without interfering data from other jails. It is also possible, to put the IP RRs also in the subdomain, when there is no need for common lists. But there must be also a combination of IP and TXT record, as most blacklist solutions expect to get an IP address as result, even when a single TXT records could be enough.

Thinking further, this could be also an core module for fail2ban, like it is working in certbot with the dns_rfc2136 extension.
As this is also an [python project](https://github.com/certbot/certbot/blob/master/certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py), implementation could be easy.

**Comment by ftasnetamot** [comment]
I am making progress. Finally I figured out, that OpenResty (an nginx-clone with several additions) is capable to do, what I want:
- accept tcp connections as stream
- check incoming ip against a dns-blocklist
- if listed, proxy to an OpenResty instance, which gives appropriate error message and logs the access (for further reading thought fail2ban)
- if not listed, forward with transparent ip to the original application, which now runs on a backend interface

There was some deep diving into the possibilities, fortunately lua is able to wrap systemcalls in this case to getHostByName() in preread.
IP-Transparent proxing is also possible.

Now I have a bash script, which provides me the possibility to set and remove dns records on my own blocklist subdomain
But I really would like, to do that pythonic. In the fail2ban code I detected, that it is possible, to load a python-module, which can do ban/unban-actions.  However I have not found an prototype, how such a module could look like.

Also, when I compare [the certbot rfc2136 module](https://github.com/certbot/certbot/blob/master/certbot-dns-rfc2136/certbot_dns_rfc2136/_internal/dns_rfc2136.py) with the [fail2ban ipdns.py module](https://github.com/fail2ban/fail2ban/blob/master/fail2ban/server/ipdns.py), I see, that rfc2136 could be just an extension with some more functions, to set and remove ip RRs on a nameserver.

Just to see, what my bash-file currently does for un-/blocking via dnsbl:
```
#!/bin/bash

### action is either "add" or "delete" ###
action=$1
IP=$2;

reverseip () {
    local IFS
    IFS=.
    set -- $1
    echo $4.$3.$2.$1
}

revIP=$(reverseip $IP)

nsupdate -k certbot.ddns.key << EOF
 zone blocklist.my-domain.top
 update ${action} ${revIP}.blocklist.my-domain.top 60 IN A 127.0.0.1
 send
 quit
EOF
```

**Comment by ftasnetamot** [comment]
Ok,  I am now one step further:  I can also collect whitelist data, a kind of reverse fail2ban, a succeed2unban:
When I track succesfull logins to pop3 or imap, I can feed them in a whitelist, which will override the early connection blocklist in exim.
That gives the opportunity, to give legit user a chance, to unblock a potentially blocked smtp server, for sending mail, just through looking for new mail.  For this to work, I need to have separate blocklists for individual services, especially I cannot block pop3/imap, because an IP address tried to spam my smtp server.
And to prevent those, who use the same ip address like regular users for spamming, I repeat the block-configuration after the user authentication for non authenticated users.  This gives a very stable setup:  I can block smtp immediately after connect for almost all offenders. However, when an IP address is recognized beeing used in the last n hours from authenticated user (via pop3/imap) I allow also authentication. If authentication fails, I check again the blocklist and disallow further commands.

Keeping verified IPs for approx 24h in the dns-list, should be a good compromise. Especially, when every positive verification is used to prolong this period. From fail2bans perspective this looks also like blocking, but it is unblocking ;-)

This can be adapted further: If you give people for example a chance, to login to an http(s) service, you can immediately set them on a welcome list and not longer blocking them. You still can block rude IPs from your webserver, if an nginx or someother instance in front of it, redirects all requests from banned ips to a verification page. Once a client verifies positive, he can do an reload the the original page, and is not longer blocked.  Much better, than just dropping packets.

Dropping packets should be the ultimate last step. And and with real large DDOS attacks, this must happen ways in front of the server, by denying routing near to the single attack nodes. But this tools are unfortunately not in reach to smaller server administrators.

**Reply by sebres** [reply]
> a kind of reverse fail2ban, a succeed2unban

It is also possible with an action, which chain would simply added before anything else in INPUT and `blocktype=RETURN`, so it whitelist the packets before other fail2ban rules. An example for similar jail is `pass2allow-ftp`.

> That gives the opportunity, to give legit user a chance, to unblock a potentially blocked smtp server, for sending mail

1. Or one simply avoid situations where a legit user gets banned.
For instance a brute-force imply that the password changes by every attempt (it makes no sense to test a user with the same password again and again), whether a legitimate user would do few typos, so uses mostly the same user name and password for login. In this case one could log the password hash (rewrite REs to capture it with `<F-PWDHASH>...</F-PWDHASH>` in filter), and then write own `ignorecommand` to write and check (IP, user, pwdhash) tuple of last X attempts to some database, so you'd ignore users having few entries and don't ban them.
2. Another possibility would be to use very small initial `bantime` and `bantime.increment = true`, so if legitimate users get banned, they will be banned for short time (at least by few first attempts). And/or increase `maxretry` and `findtime`.
3. Alternative solution is to provide user a possibility in WEB-UI to unban himself (e. g. URI with a request to some backend, invoking `fail2ban-client unban "$ip"` or logging to file monitored by fail2ban to whitelist them in pass2allow style so unban it via action doing `nohup fail2ban-client unban "$ip" &` internally)...

There are many possibilities basically, but the best will be always to avoid false positives.
